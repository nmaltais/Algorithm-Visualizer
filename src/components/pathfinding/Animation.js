/* eslint-disable no-param-reassign */

function updatedGrid(grid, pos, newCell) {
  const gridCopy = [...grid];
  gridCopy[pos.y][pos.x] = { ...gridCopy[pos.y][pos.x], ...newCell };
  return gridCopy;
}

export function playAnimation(grid, setGrid, animationQueue, currentStep, lastRequestID, timeOut, speedPercentageRef, setAnimationState) {
  setAnimationState('playing');

  function displayNextStep() {
    const { steps, type } = animationQueue.current[0];
    const { x, y } = steps[currentStep.current];

    if (grid[y][x] !== undefined) {
      if (type === 'path') setGrid(updatedGrid(grid, { y, x }, { visited: true, onPath: true }));
      else if (type === 'visits') {
        if (currentStep.current > 0) {
          setGrid(updatedGrid(grid, { y: steps[currentStep.current - 1].y, x: steps[currentStep.current - 1].x }, { visited: true, visiting: false }));
        }
        setGrid(updatedGrid(grid, { y, x }, { visited: false, visiting: true }));
      }
    }

    currentStep.current += 1;
    if (currentStep.current < steps.length) {
      timeOut.current = setTimeout(() => playAnimation(grid,
        setGrid,
        animationQueue,
        currentStep,
        lastRequestID,
        timeOut,
        speedPercentageRef,
        setAnimationState),
      -20 * speedPercentageRef + 2000);
    } else {
      currentStep.current = 0;
      if (animationQueue.current.length > 1) {
        animationQueue.current.shift();
        currentStep.current = 0;
        playAnimation(grid, setGrid, animationQueue, currentStep, lastRequestID, timeOut, speedPercentageRef, setAnimationState);
      } else setAnimationState('done');
    }
  }
  // Update AnimationFrame RequestID
  lastRequestID = requestAnimationFrame(displayNextStep);
}

export function pauseAnimation(timeOut, lastRequestID, setAnimationState) {
  // Stop Animation
  cancelAnimationFrame(lastRequestID.current);
  clearTimeout(timeOut.current);
  setAnimationState('paused');
}
