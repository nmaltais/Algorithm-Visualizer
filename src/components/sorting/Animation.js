/* eslint-disable no-param-reassign */

const defaultColor = 'rgb(116, 172, 255)';
const compareColor1 = '#ff7961';
const compareColor2 = '#f44336';

export function playAnimation(currentStep, lastCompared, lastSwapped, lastCursor1Pos, lastCursor2Pos, lastCursor3Pos, lastRequestID,
  bars, steps, timeOut, speedPercentageRef, setAnimationState) {
  setAnimationState('playing');

  function displayNextStep() {
    const step = steps[currentStep.current];

    switch (step.action) {
      case 'compare':
        if (bars) {
          if (lastCompared.current.length > 0) { // Reset colors for last compared columns
            bars[lastCompared.current[0]].ref.current.style.backgroundColor = defaultColor;
            bars[lastCompared.current[1]].ref.current.style.backgroundColor = defaultColor;
          }
          // Update colors for currently compared columns
          bars[step.i].ref.current.style.backgroundColor = compareColor1;
          bars[step.j].ref.current.style.backgroundColor = compareColor2;

          lastCompared.current = [step.i, step.j];
        }

        break;
      case 'swap': {
        if (bars) {
          // Swap heights, colors, and cursor for currently swapped indexes (giving the illusion that the bars being swapped)
          const iHeight = bars[step.i].ref.current.querySelector('.bar').style.height;
          const jHeight = bars[step.j].ref.current.querySelector('.bar').style.height;

          const iColor = bars[step.i].ref.current.style.backgroundColor;
          const jColor = bars[step.j].ref.current.style.backgroundColor;
          // Swap values
          bars[step.i].ref.current.querySelector('.barValue').innerText = parseInt(jHeight, 10);
          bars[step.j].ref.current.querySelector('.barValue').innerText = parseInt(iHeight, 10);
          // Swap Heights
          bars[step.i].ref.current.querySelector('.bar').style.height = jHeight;
          bars[step.j].ref.current.querySelector('.bar').style.height = iHeight;
          // Swap colors
          bars[step.i].ref.current.style.backgroundColor = jColor;
          bars[step.j].ref.current.style.backgroundColor = iColor;
          lastSwapped.current = [step.i, step.j];
        }
        break;
      }
      case 'move cursor1':
        if (bars) {
          if (lastCursor1Pos.current !== null) {
            bars[lastCursor1Pos.current].ref.current.querySelector('.cursor1').style.display = 'none';
          }

          if (step.pos >= bars.length) {
            lastCursor1Pos.current = null;
          } else {
            bars[step.pos].ref.current.querySelector('.cursor1').style.display = 'block';
            if (step.label) bars[step.pos].ref.current.querySelector('.cursor1').innerText = step.label;
            lastCursor1Pos.current = step.pos;
          }

          if (lastCompared.current.length > 0) { // Reset colors for last compared columns
            bars[lastCompared.current[0]].ref.current.style.backgroundColor = defaultColor;
            bars[lastCompared.current[1]].ref.current.style.backgroundColor = defaultColor;
          }
          if (lastSwapped.current.length > 0) { // Reset colors for last swapped columns
            bars[lastSwapped.current[0]].ref.current.style.backgroundColor = defaultColor;
            bars[lastSwapped.current[1]].ref.current.style.backgroundColor = defaultColor;
          }
        }
        break;
      case 'move cursor2':
        if (bars) {
          if (lastCursor2Pos.current !== null) {
            bars[lastCursor2Pos.current].ref.current.querySelector('.cursor2').style.display = 'none';
          }

          if (step.pos >= bars.length) {
            lastCursor2Pos.current = null;
          } else {
            bars[step.pos].ref.current.querySelector('.cursor2').style.display = 'block';
            if (step.label) bars[step.pos].ref.current.querySelector('.cursor2').innerText = step.label;
            lastCursor2Pos.current = step.pos;
          }

          if (lastCompared.current.length > 0) { // Reset colors for last compared columns
            bars[lastCompared.current[0]].ref.current.style.backgroundColor = defaultColor;
            bars[lastCompared.current[1]].ref.current.style.backgroundColor = defaultColor;
          }
          if (lastSwapped.current.length > 0) { // Reset colors for last swapped columns
            bars[lastSwapped.current[0]].ref.current.style.backgroundColor = defaultColor;
            bars[lastSwapped.current[1]].ref.current.style.backgroundColor = defaultColor;
          }
        }
        break;
      case 'move cursor3':
        if (bars) {
          if (lastCursor3Pos.current !== null) {
            bars[lastCursor3Pos.current].ref.current.querySelector('.cursor3').style.display = 'none';
          }

          if (step.pos >= bars.length) {
            lastCursor3Pos.current = null;
          } else {
            bars[step.pos].ref.current.querySelector('.cursor3').style.display = 'block';
            if (step.label) bars[step.pos].ref.current.querySelector('.cursor3').innerText = step.label;
            lastCursor3Pos.current = step.pos;
          }

          if (lastCompared.current.length > 0) { // Reset colors for last compared columns
            bars[lastCompared.current[0]].ref.current.style.backgroundColor = defaultColor;
            bars[lastCompared.current[1]].ref.current.style.backgroundColor = defaultColor;
          }
          if (lastSwapped.current.length > 0) { // Reset colors for last swapped columns
            bars[lastSwapped.current[0]].ref.current.style.backgroundColor = defaultColor;
            bars[lastSwapped.current[1]].ref.current.style.backgroundColor = defaultColor;
          }
        }
        break;
      default:
        console.log('default case');
        break;
    }
    currentStep.current += 1;
    if (currentStep.current < steps.length) {
      timeOut.current = setTimeout(() => playAnimation(currentStep,
        lastCompared,
        lastSwapped,
        lastCursor1Pos,
        lastCursor2Pos,
        lastCursor3Pos,
        lastRequestID,
        bars,
        steps,
        timeOut,
        speedPercentageRef,
        setAnimationState),
      -20 * speedPercentageRef.current + 2000);
    } else {
      setAnimationState('done');
    }
  }
  // Update AnimationFrame RequestID
  lastRequestID.current = requestAnimationFrame(displayNextStep);
}


export function pauseAnimation(timeOut, lastRequestID, setAnimationState) {
  // Stop Animation
  cancelAnimationFrame(lastRequestID.current);
  clearTimeout(timeOut.current);
  setAnimationState('paused');
}

function resetbarRefs(bars, list) {
  list.forEach((num, i) => {
    bars[i].ref.current.style.backgroundColor = defaultColor;
    bars[i].ref.current.querySelector('.bar').style.height = `${num}vh`;
    bars[i].ref.current.querySelector('.barValue').innerText = `${num}`;
    bars[i].ref.current.querySelector('.cursor1').style.display = 'none';
    bars[i].ref.current.querySelector('.cursor2').style.display = 'none';
    bars[i].ref.current.querySelector('.cursor3').style.display = 'none';
  });
}
export function resetAnimation(currentStep, timeOut, lastRequestID, bars, list, setAnimationState) {
  setAnimationState('init');
  // Stop Animation
  cancelAnimationFrame(lastRequestID.current);
  clearTimeout(timeOut.current);
  // Reset currentStep and bars
  currentStep.current = 0;
  if (bars.length > 0 && bars[0].ref.current) resetbarRefs(bars, list);
}
