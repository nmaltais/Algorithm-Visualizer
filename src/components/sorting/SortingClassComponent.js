/* eslint-disable react/destructuring-assignment */
/* eslint-disable no-param-reassign */
/* eslint-disable no-plusplus */
import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { Button, Grid } from 'semantic-ui-react';
import './Sorting.scss';
import { BubbleSort, InsertionSort } from './sorting-algorithms';
import { playAnimation, pauseAnimation, resetAnimation } from './Animation';


class SortingClassComponent extends Component {
  constructor(props) {
    super(props);
    this.state = {
      list: [20, 64, 23, 35, 35, 75, 34, 69,
        76, 57, 9, 67, 33, 58, 7, 27, 22,
        56, 37, 5, 79, 36, 46, 65],
      info: '',
      steps: [],
      animationState: 'init',
      speedPercentage: 80,
    };
    this.barRefs = React.createRef();
    this.currentStep = React.createRef();
    this.lastCompared = React.createRef();
    this.lastCursorPos = React.createRef();
    this.lastRequestID = React.createRef();
    this.timeOut = React.createRef();
    this.speedPercentageRef = React.createRef();
  }

  componentDidMount() {
    const { match } = this.props;

    this.barRefs.current = [];
    this.currentStep.current = 0;
    this.lastCompared.current = [];
    this.lastCursorPos.current = null;
    this.lastRequestID.current = null;
    this.timeOut.current = null;
    this.speedPercentageRef.current = 80;

    const alg = match.params.alg ? match.params.alg : 'bubble-sort';
    this.getAlgOutput(alg);
    this.createBarsFromList();
  }

  componentDidUpdate(prevProps) {
    if (this.props.match.params.alg !== prevProps.match.params.alg) {
      console.log('Route change!');
      this.getAlgOutput(this.props.match.params.alg);
      resetAnimation(this.currentStep, this.timeOut, this.lastRequestID, this.barRefs, this.state.list, this.setAnimationState);
    }
  }

  getAlgOutput = (path) => {
    let algOutput = null;

    switch (path) {
      case 'bubble-sort':
        algOutput = BubbleSort(this.state.list.slice());
        break;
      case 'insertion-sort':
        algOutput = InsertionSort(this.state.list.slice());
        break;
      default:
        console.log('no-alg');
        break;
    }
    if (algOutput !== null) {
      this.setState({ steps: algOutput.steps });
      this.setState({ info: algOutput.info });
    }
  }

  createBarsFromList = () => {
    const { list } = this.state;
    const localBars = list.map((num, i) => {
      const ref = React.createRef();
      const bar = (
        <span
          key={i}
          ref={ref}
          style={{ height: `${num}vh`, width: `${((100 * (12 / 16)) - 20) / this.state.list.length}vw`, maxWidth: '50px' }}
          className="bar"
        >
          {num}
        </span>
      );
      this.barRefs.current.push(ref);
      return bar;
    });
    this.setState({ bars: localBars });
  }

  setAnimationState = (animationState) => {
    this.setState({ animationState });
  }

  changeSpeed = (e) => {
    this.speedPercentageRef.current = e.target.value;
    this.setState({ speedPercentage: e.target.value });
  }


  render() {
    return (
      <Grid stackable id="container">
        <Grid.Row>
          <Grid.Column width={3}>
            <h1>{this.state.info.title}</h1>
            <p>
              <span>{'Worst Time: '}</span>
              <span>{this.state.info.worstTime}</span>
              <br />
              <span>{'Avg. Time: '}</span>
              <span>{this.state.info.avgTime}</span>
              <br />
              <span>{'Best Time: '}</span>
              <span>{this.state.info.bestTime}</span>
              <br />
              <span>{'Space: '}</span>
              <span>{this.state.info.space}</span>
            </p>
          </Grid.Column>
          <Grid.Column width={13}>
            {this.state.animationState === 'playing'
              ? <Button id="pauseBtn" icon="pause" content="Pause" onClick={() => pauseAnimation(this.timeOut, this.lastRequestID, this.setAnimationState)} />
              : (
                <Button
                  id="playBtn"
                  disabled={this.state.animationState === 'done'}
                  icon="play"
                  content="Play"
                  onClick={() => playAnimation(this.currentStep, this.lastCompared, this.lastCursorPos, this.lastRequestID,
                    this.barRefs, this.state.steps, this.timeOut, this.speedPercentageRef,
                    this.setAnimationState)}
                />
              )}
            <Button
              id="resetBtn"
              icon="repeat"
              content="Reset"
              onClick={() => resetAnimation(this.currentStep, this.timeOut, this.lastRequestID, this.barRefs, this.state.list, this.setAnimationState)}
            />
            {' Speed: '}
            <input
              id="speedSlider"
              type="range"
              min={0}
              max={100}
              value={this.state.speedPercentage}
              onChange={this.changeSpeed}
            />
            {`${this.state.speedPercentage}%`}
            <br />
            <br />
            {this.state.bars}
          </Grid.Column>
        </Grid.Row>
      </Grid>
    );
  }
}

export default withRouter(SortingClassComponent);
