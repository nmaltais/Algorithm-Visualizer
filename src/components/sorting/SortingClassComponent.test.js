import React from 'react';
import { shallow, mount } from 'enzyme';
import SortingClassComponent from './SortingClassComponent';
import * as Animation from './Animation';


afterEach(() => {
  jest.clearAllMocks();
});

describe('SortingClassComponent', () => {
  it('renders bubble-sort correctly', () => {
    const wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'bubble-sort' } }} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('defaults to bubble-sort', () => {
    const wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'something' } }} />);
    const mockGetAlgOutput = jest.spyOn(wrapper.instance(), 'getAlgOutput');
    wrapper.setProps({ match: { params: { } } });
    expect(mockGetAlgOutput).toHaveBeenCalledWith('bubble-sort');
    wrapper.setProps({ match: { } });
    expect(mockGetAlgOutput).toHaveBeenCalledWith('bubble-sort');
  });

  it('should display bubble-sort if no valid path selected', () => {
    let wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: { alg: '' } }} />);
    expect(wrapper.contains(<h1>Bubble Sort</h1>)).toBeTruthy();
    wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: {} }} />);
    expect(wrapper.contains(<h1>Bubble Sort</h1>)).toBeTruthy();
    wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'qwerty' } }} />);
    expect(wrapper.contains(<h1>Bubble Sort</h1>)).toBeTruthy();
  });

  it('switches from play to pause and vice-versa', () => {
    const wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'bubble-sort' } }} />);
    wrapper.find('#playBtn').simulate('click');
    expect(wrapper.find('#pauseBtn').exists()).toBeTruthy();
    expect(!wrapper.find('#playBtn').exists()).toBeTruthy();
    wrapper.find('#pauseBtn').simulate('click');
    expect(wrapper.find('#playBtn').exists()).toBeTruthy();
    expect(!wrapper.find('#pauseBtn').exists()).toBeTruthy();
  });

  it('Displays play btn after hitting reset', () => {
    const wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'bubble-sort' } }} />);
    wrapper.find('#playBtn').simulate('click');
    wrapper.find('#resetBtn').simulate('click');
    expect(wrapper.find('#playBtn').exists()).toBeTruthy();
    expect(!wrapper.find('#pauseBtn').exists()).toBeTruthy();
  });

  it('playBtn click calls playAnimation', () => {
    const mockPlayAnimation = jest.spyOn(Animation, 'playAnimation');
    const wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'bubble-sort' } }} />);
    wrapper.find('#playBtn').simulate('click');
    expect(mockPlayAnimation).toHaveBeenCalled();
    wrapper.find('#resetBtn').simulate('click');
  });
  it('pauseBtn click calls pauseAnimation', () => {
    const mockPauseAnimation = jest.spyOn(Animation, 'pauseAnimation');
    const wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'bubble-sort' } }} />);
    wrapper.find('#playBtn').simulate('click'); // To make pauseBtn appear
    wrapper.find('#pauseBtn').simulate('click');
    expect(mockPauseAnimation).toHaveBeenCalled();
  });
  it('resetBtn click calls resetAnimation', () => {
    const mockResetAnimation = jest.spyOn(Animation, 'resetAnimation');
    const wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'bubble-sort' } }} />);
    wrapper.find('#resetBtn').simulate('click');
    expect(mockResetAnimation).toHaveBeenCalled();
  });
  it('animationSpeed slider change calls changeSpeed', () => {
    const wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'bubble-sort' } }} />);
    const mockChangeSpeed = jest.spyOn(wrapper.instance(), 'changeSpeed');
    wrapper.instance().forceUpdate();
    wrapper.find('#speedSlider').simulate('change', { target: { value: 69 } });
    expect(mockChangeSpeed).toHaveBeenCalled();
  });
  it('animationSpeed slider changes the state correctly', () => {
    const wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'bubble-sort' } }} />);
    expect(wrapper.instance().state.speedPercentage).toEqual(100);
    expect(wrapper.instance().speedPercentageRef.current).toEqual(100);
    wrapper.find('#speedSlider').simulate('change', { target: { value: 20 } });
    expect(wrapper.instance().state.speedPercentage).toEqual(20);
    expect(wrapper.instance().speedPercentageRef.current).toEqual(20);
    wrapper.find('#speedSlider').simulate('change', { target: { value: 78 } });
    expect(wrapper.instance().state.speedPercentage).toEqual(78);
    expect(wrapper.instance().speedPercentageRef.current).toEqual(78);
  });
  it('listLengthSlider slider change calls generateRandomBars', () => {
    const wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'merge-sort' } }} />);
    const mockGenerateRandomBars = jest.spyOn(wrapper.instance(), 'generateRandomBars');
    wrapper.instance().forceUpdate();
    wrapper.find('#listLengthSlider').simulate('change', { target: { value: 40 } });
    expect(mockGenerateRandomBars).toHaveBeenCalled();
  });
  it('listLengthSlider slider changes the state & bars correctly', () => {
    const wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'insertion-sort' } }} />);
    wrapper.find('#listLengthSlider').simulate('change', { target: { value: 40 } });
    expect(wrapper.instance().state.listLength).toEqual(40);
    expect(wrapper.instance().state.bars).toHaveLength(40);
    wrapper.find('#listLengthSlider').simulate('change', { target: { value: 3 } });
    expect(wrapper.instance().state.listLength).toEqual(3);
    expect(wrapper.instance().state.bars).toHaveLength(3);
  });
  it('animation btns trigger correct state changes', () => {
    const wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'bubble-sort' } }} />);
    expect(wrapper.instance().state.animationState).toEqual('init');
    wrapper.find('#playBtn').simulate('click');
    expect(wrapper.instance().state.animationState).toEqual('playing');
    wrapper.find('#pauseBtn').simulate('click');
    expect(wrapper.instance().state.animationState).toEqual('paused');
    wrapper.find('#playBtn').simulate('click');
    expect(wrapper.instance().state.animationState).toEqual('playing');
    wrapper.find('#resetBtn').simulate('click');
    expect(wrapper.instance().state.animationState).toEqual('init');
  });
  it('Changing url params resets the animation', () => {
    const wrapper = shallow(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'bubble-sort' } }} />);
    expect(wrapper.instance().state.animationState).toEqual('init');
    wrapper.find('#playBtn').simulate('click');
    expect(wrapper.instance().state.animationState).toEqual('playing');
    wrapper.setProps({ match: { params: { alg: 'insertion-sort' } } });
    const mockResetAnimation = jest.spyOn(Animation, 'resetAnimation');
    expect(mockResetAnimation).toHaveBeenCalled();
    expect(wrapper.instance().state.animationState).toEqual('init');
  });
});

describe('Animation', () => {
  it('Updates step count as animation plays', async () => {
    const mockPlayAnimation = jest.spyOn(Animation, 'playAnimation');
    const mockRequestAnimationFrame = jest.spyOn(global, 'requestAnimationFrame');
    const wrapper = mount(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'quick-sort' } }} />);
    wrapper.instance().speedPercentageRef.current = 90;

    expect(wrapper.instance().currentStep.current).toEqual(0);
    wrapper.find('#playBtn').first().simulate('click');
    expect(mockRequestAnimationFrame).toHaveBeenCalledWith(expect.any(Function));

    // Let the animation play for 5 steps and make sure 5 frames have been requested.
    await new Promise((r) => setTimeout(r, (-20 * wrapper.instance().speedPercentageRef.current + 2000) * 5));
    expect(mockPlayAnimation).toHaveBeenCalledTimes(1);
    expect(mockRequestAnimationFrame).toHaveBeenCalledTimes(5);
    expect(wrapper.instance().currentStep.current).toEqual(5);

    // Hack: Stop animation so it doesn't conflict with next test
    wrapper.find('#resetBtn').first().simulate('click');
  });

  it('Resets the bars', async () => {
    const mockRequestAnimationFrame = jest.spyOn(global, 'requestAnimationFrame');
    const wrapper = mount(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'insertion-sort' } }} />);
    wrapper.instance().speedPercentageRef.current = 90;

    // Let the animation play for 5 steps.
    wrapper.find('#playBtn').last().simulate('click');
    await new Promise((r) => setTimeout(r, (-20 * wrapper.instance().speedPercentageRef.current + 2000) * 5));
    expect(mockRequestAnimationFrame).toHaveBeenCalledTimes(5);
    expect(wrapper.instance().currentStep.current).toEqual(5);

    // Press reset button and make sure the bars are reset.
    wrapper.find('#resetBtn').first().simulate('click');
    wrapper.instance().state.list.forEach((num, i) => {
      expect(wrapper.instance().state.bars[i].ref.current.style.backgroundColor).toEqual('rgb(116, 172, 255)');
      expect(wrapper.instance().state.bars[i].ref.current.querySelector('.bar').style.height).toEqual(`${num}vh`);
      expect(wrapper.instance().state.bars[i].ref.current.querySelector('.barValue').innerText).toEqual(`${num}`);
      expect(wrapper.instance().state.bars[i].ref.current.querySelector('.cursor1').style.display).toEqual('none');
      expect(wrapper.instance().state.bars[i].ref.current.querySelector('.cursor2').style.display).toEqual('none');
      expect(wrapper.instance().state.bars[i].ref.current.querySelector('.cursor3').style.display).toEqual('none');
    });

    // Hack: Stop animation so it doesn't conflict with next test
    wrapper.find('#resetBtn').first().simulate('click');
  });

  it('Generates random bars', async () => {
    const wrapper = mount(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'merge-sort' } }} />);

    const barsBefore = wrapper.instance().state.bars;
    wrapper.find('#newRandomSet').last().simulate('click');
    const barsAfter = wrapper.instance().state.bars;
    expect(barsAfter).toHaveLength(barsBefore.length);
    expect(barsAfter).not.toEqual(barsBefore);
  });

  it('Pauses the animation', async () => {
    jest.clearAllMocks();
    const mockRequestAnimationFrame = jest.spyOn(global, 'requestAnimationFrame');
    const wrapper = mount(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'bubble-sort' } }} />);
    wrapper.instance().speedPercentageRef.current = 90;

    // Let the animation play for 5 steps
    wrapper.find('#playBtn').first().simulate('click');
    await new Promise((r) => setTimeout(r, (-20 * wrapper.instance().speedPercentageRef.current + 2000) * 5));
    expect(mockRequestAnimationFrame).toHaveBeenCalledTimes(5);
    expect(wrapper.instance().currentStep.current).toEqual(5);

    // Pause the animation, wait and make sure no more frames having been requested.
    wrapper.find('#pauseBtn').first().simulate('click');
    await new Promise((r) => setTimeout(r, (-20 * wrapper.instance().speedPercentageRef.current + 2000) * 5));
    expect(mockRequestAnimationFrame).toHaveBeenCalledTimes(5);
    expect(wrapper.instance().currentStep.current).toEqual(5);

    // Hack: Stop animation so it doesn't conflict with next test
    wrapper.find('#resetBtn').first().simulate('click');
  });

  it('Sets State to "done" when animation is over', async () => {
    jest.clearAllMocks();
    const wrapper = mount(<SortingClassComponent.WrappedComponent match={{ params: { alg: 'selection-sort' } }} />);
    wrapper.instance().speedPercentageRef.current = 80;
    wrapper.instance().state.list = [3, 2, 5];
    wrapper.instance().getAlgOutput('selection-sort');

    expect(wrapper.instance().state.animationState).toEqual('init');
    wrapper.find('#playBtn').first().simulate('click');
    expect(wrapper.instance().state.animationState).toEqual('playing');
    await new Promise((r) => setTimeout(r, (-20 * wrapper.instance().speedPercentageRef.current + 2000) * wrapper.instance().state.steps.length));
    expect(wrapper.instance().state.animationState).toEqual('done');

    // Hack: Stop animation so it doesn't conflict with next test
    wrapper.find('#resetBtn').first().simulate('click');
  });
});
