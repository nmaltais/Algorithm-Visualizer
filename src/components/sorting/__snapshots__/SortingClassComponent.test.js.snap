// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortingClassComponent renders bubble-sort correctly 1`] = `
<Grid
  id="container"
  stackable={true}
>
  <GridRow>
    <GridColumn
      width={3}
    >
      <h1>
        Bubble Sort
      </h1>
      <p>
        <span>
          Worst Time: 
        </span>
        <span>
          O(n^2)
        </span>
        <br />
        <span>
          Avg. Time: 
        </span>
        <span>
          O(n^2)
        </span>
        <br />
        <span>
          Best Time: 
        </span>
        <span>
          O(n)
        </span>
        <br />
        <span>
          Space: 
        </span>
        <span>
          O(n)
        </span>
      </p>
    </GridColumn>
    <GridColumn
      width={13}
    >
      <Button
        as="button"
        content="Play"
        disabled={false}
        icon="play"
        id="playBtn"
        onClick={[Function]}
      />
      <Button
        as="button"
        content="Reset"
        icon="repeat"
        id="resetBtn"
        onClick={[Function]}
      />
       Speed: 
      <input
        id="speedSlider"
        max={100}
        min={0}
        onChange={[Function]}
        type="range"
        value={80}
      />
      80%
      <br />
      <br />
      <span
        className="bar"
        key="0"
        style={
          Object {
            "height": "20vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        20
      </span>
      <span
        className="bar"
        key="1"
        style={
          Object {
            "height": "64vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        64
      </span>
      <span
        className="bar"
        key="2"
        style={
          Object {
            "height": "23vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        23
      </span>
      <span
        className="bar"
        key="3"
        style={
          Object {
            "height": "35vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        35
      </span>
      <span
        className="bar"
        key="4"
        style={
          Object {
            "height": "35vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        35
      </span>
      <span
        className="bar"
        key="5"
        style={
          Object {
            "height": "75vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        75
      </span>
      <span
        className="bar"
        key="6"
        style={
          Object {
            "height": "34vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        34
      </span>
      <span
        className="bar"
        key="7"
        style={
          Object {
            "height": "69vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        69
      </span>
      <span
        className="bar"
        key="8"
        style={
          Object {
            "height": "76vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        76
      </span>
      <span
        className="bar"
        key="9"
        style={
          Object {
            "height": "57vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        57
      </span>
      <span
        className="bar"
        key="10"
        style={
          Object {
            "height": "9vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        9
      </span>
      <span
        className="bar"
        key="11"
        style={
          Object {
            "height": "67vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        67
      </span>
      <span
        className="bar"
        key="12"
        style={
          Object {
            "height": "33vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        33
      </span>
      <span
        className="bar"
        key="13"
        style={
          Object {
            "height": "58vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        58
      </span>
      <span
        className="bar"
        key="14"
        style={
          Object {
            "height": "7vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        7
      </span>
      <span
        className="bar"
        key="15"
        style={
          Object {
            "height": "27vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        27
      </span>
      <span
        className="bar"
        key="16"
        style={
          Object {
            "height": "22vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        22
      </span>
      <span
        className="bar"
        key="17"
        style={
          Object {
            "height": "56vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        56
      </span>
      <span
        className="bar"
        key="18"
        style={
          Object {
            "height": "37vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        37
      </span>
      <span
        className="bar"
        key="19"
        style={
          Object {
            "height": "5vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        5
      </span>
      <span
        className="bar"
        key="20"
        style={
          Object {
            "height": "79vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        79
      </span>
      <span
        className="bar"
        key="21"
        style={
          Object {
            "height": "36vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        36
      </span>
      <span
        className="bar"
        key="22"
        style={
          Object {
            "height": "46vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        46
      </span>
      <span
        className="bar"
        key="23"
        style={
          Object {
            "height": "65vh",
            "maxWidth": "50px",
            "width": "2.2916666666666665vw",
          }
        }
      >
        65
      </span>
    </GridColumn>
  </GridRow>
</Grid>
`;
